limacharlie/artifact.go

```diff
@@ -76,6 +82,125 @@ func (org Organization) ArtifactRuleDelete(ruleName ArtifactRuleName) error {
+}
+
+func (org Organization) UploadArtifact(data io.Reader, size int64, hint string, source string, artifactId string, originalPath string, nDaysRetention int, ingestion_key string) error {
+
+	// Assemble headers
+	headers := map[string]string{}
+	headers["lc-source"] = source
+	headers["lc-hint"] = hint
+	if artifactId != "" {
+		headers["lc-artifact-id"] = artifactId
+	} else {
+		headers["lc-artifact-id"] = uuid.New().String()
+	}
+	if originalPath != "" {
+		absolutePath, _ := filepath.Abs(originalPath)
+		headers["lc-path"] = base64.StdEncoding.EncodeToString([]byte(absolutePath))
+	}
+	headers["lc-retention-days"] = fmt.Sprint(nDaysRetention)
+
+	// Get artifacts URL
+	urls, err := org.GetURLs()
+	if err != nil {
+		return fmt.Errorf("failed resolving org URLs: %v", err)
+	}
+	uploadUrl, ok := urls["artifacts"]
+	if !ok {
+		return errors.New("artifacts URL not found in org URLs")
+	}
+
+	fileSize := size
+
+	// Build request
+	combined := fmt.Sprintf("%s:%s", org.GetOID(), ingestion_key)
+	base64_encoded_ingestion_key := base64.StdEncoding.EncodeToString([]byte(combined))
+
+	if MAX_UPLOAD_PART_SIZE > fileSize {
+		// Simple single-chunk upload.
+		headers = headers
+
+		c := &http.Client{}
+
+		req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("https://%s/ingest", uploadUrl), data)
+		req.Header.Set("Content-Type", "application/octet-stream")
+		req.Header.Set("Authorization", fmt.Sprintf("Basic %s", base64_encoded_ingestion_key))
+		if err != nil {
+			return err
+		}
+
+		for k, v := range headers {
+			req.Header.Set(k, v)
+		}
+
+		httpResp, err := c.Do(req)
+		if err != nil {
+			return err
+		}
+		if httpResp.StatusCode != 200 {
+			return fmt.Errorf("failed to POST artifact, http status: %d", httpResp.StatusCode)
+		}
+
+		return nil
+
+	} else {
+		// Multi-part upload.
+		partId := 0
+		if artifactId == "" {
+			headers["lc-artifact-id"] = uuid.New().String()
+		}
+
+		for {
+			chunk := make([]byte, MAX_UPLOAD_PART_SIZE)
+			n, err := data.Read(chunk)
+			if err != nil && err != io.EOF {
+				return err
+			}
+			if n == 0 {
+				break
+			}
+			chunk = chunk[:n]
+			readerChunk := bytes.NewReader(chunk)
+
+			if n < MAX_UPLOAD_PART_SIZE {
+				headers["lc-part"] = "done"
+			} else {
+				headers["lc-part"] = fmt.Sprintf("%d", partId)
+			}
+
+			c := &http.Client{}
+			data = readerChunk
+
+			req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("https://%s/ingest", uploadUrl), data)
+			req.Header.Set("Content-Type", "application/octet-stream")
+			req.Header.Set("Authorization", fmt.Sprintf("Basic %s", base64_encoded_ingestion_key))
+			if err != nil {
+				return err
+			}
+
+			for k, v := range headers {
+				req.Header.Set(k, v)
+			}
+
+			httpResp, err := c.Do(req)
+			if err != nil {
+				return err
+			}
+			if httpResp.StatusCode != 200 {
+				return fmt.Errorf("failed to POST artifact, http status: %d", httpResp.StatusCode)
+			}
+
+			partId++
+		}
+	}
+	return err
```

Comment:
```suggestion
	// Get artifacts URL
	urls, err := org.GetURLs()
	if err != nil {
		return fmt.Errorf("failed resolving org URLs: %v", err)
	}
	uploadUrl, ok := urls["artifacts"]
	if !ok {
		return errors.New("artifacts URL not found in org URLs")
	}
	reqUrl := fmt.Sprintf("https://%s/ingest", uploadUrl)

	if artifactId == "" {
		headers["lc-artifact-id"] = uuid.New().String()
	}

	// Build request
	combined := fmt.Sprintf("%s:%s", org.GetOID(), ingestion_key)
	creds := base64.StdEncoding.EncodeToString([]byte(combined))

	c := &http.Client{}
	defer c.CloseIdleConnections()
	partId := 0
	endOffset := int64(0)
	for {
		// Read from the data in chunks of MAX_UPLOAD_PART_SIZE so we can
		// upload in parts if the file is too big.
		chunk := make([]byte, MAX_UPLOAD_PART_SIZE)
		n, err := data.Read(chunk)
		if err != nil && err != io.EOF {
			return err
		}
		if n == 0 {
			break
		}
		chunk = chunk[:n]

		// If this is the last chunk, set the part to "done".
		endOffset += int64(n)
		if endOffset > size {
			return fmt.Errorf("got more data (%d bytes) than expected (%d bytes)", endOffset, size)
		}
		if endOffset == size {
			headers["lc-part"] = "done"
		} else {
			headers["lc-part"] = fmt.Sprintf("%d", partId)
		}

		// Prepare the request.
		req, err := http.NewRequest(http.MethodPost, reqUrl, bytes.NewBuffer(chunk))
		if err != nil {
			return err
		}
		req.Header.Set("Content-Type", "application/octet-stream")
		req.Header.Set("Authorization", fmt.Sprintf("Basic %s", creds))

		// Add the dynamic headers.
		for k, v := range headers {
			req.Header.Set(k, v)
		}

		// Send the request.
		httpResp, err := c.Do(req)
		if err != nil {
			return err
		}

		// Check if the API liked it.
		if httpResp.StatusCode != 200 {
			return fmt.Errorf("failed to POST artifact, http status: %d", httpResp.StatusCode)
		}
		partId++
	}
	return nil
```

There was a few issues with the part number and offset. Here's a suggested simplification to the code.